File: 79.单词搜索.cpp
/*
 * @lc app=leetcode.cn id=79 lang=cpp
 * @lcpr version=30203
 *
 * [79] 单词搜索
 */

// @lc code=start
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        int m = board.size(), n = board[0].size();
        visited.resize(m, vector<bool>(n, false));
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == word[0]) {
                    if (backtrack(board, i, j, word, 0)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

private:
    vector<vector<bool>> visited;
    bool backtrack(vector<vector<char>>& board, int i, int j, string& word, int idx) {
        if (i >= board.size() || j >= board[0].size() || i < 0 || j < 0) {
            return false;
        }
        if (idx >= word.size()) {
            return false;
        }
        if (visited[i][j]) {
            return false;
        }
        if (board[i][j] != word[idx]) {
            return false;
        }
        if (idx == word.size() - 1) {
            return true;
        }
        visited[i][j] = true;
        if (
            backtrack(board, i + 1, j, word, idx + 1) ||
            backtrack(board, i - 1, j, word, idx + 1) ||
            backtrack(board, i, j + 1, word, idx + 1) ||
            backtrack(board, i, j - 1, word, idx + 1)
        ) {
            return true;
        }
        visited[i][j] = false;
        return false;
    }
};
// @lc code=end



/*
// @lcpr case=start
// [['A','B','C','E'],['S','F','C','S'],['A','D','E','E']]\n"ABCCED"\n
// @lcpr case=end

// @lcpr case=start
// [['A','B','C','E'],['S','F','C','S'],['A','D','E','E']]\n"SEE"\n
// @lcpr case=end

// @lcpr case=start
// [['A','B','C','E'],['S','F','C','S'],['A','D','E','E']]\n"ABCB"\n
// @lcpr case=end

 */

File: 79.单词搜索.java
/*
 * @lc app=leetcode.cn id=79 lang=java
 * @lcpr version=30305
 *
 * [79] 单词搜索
 */

// @lc code=start
class Solution {
    private boolean found = false;
    public boolean exist(char[][] board, String word) {
        if (word.length() == 0) {
            return true;
        }
        int m = board.length, n = board[0].length;
        boolean[][] visited = new boolean[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == word.charAt(0)) {
                    // visited[i][j] = true;
                    backtrack(board, visited, word, i, j, 0);
                    // visited[i][j] = false;
                }
                if (found) {
                    return true;
                }
            }
        }
        return false;
    }

    private void backtrack(char[][] board, boolean[][] visited, String word, int i, int j, int idx) {
        if (idx == word.length()) {
            found = true;
            return;
        }
        if (i < 0 || j < 0 
            || i >= board.length || j >= board[0].length 
        ) {
            return;
        }
        if (found || visited[i][j]) {
            return;
        }
        if (board[i][j] != word.charAt(idx)) {
            return;
        }
        visited[i][j] = true;
        backtrack(board, visited, word, i + 1, j, idx + 1);
        backtrack(board, visited, word, i, j + 1, idx + 1);
        backtrack(board, visited, word, i - 1, j, idx + 1);
        backtrack(board, visited, word, i, j - 1, idx + 1);
        visited[i][j] = false;
    }
}
// @lc code=end



/*
// @lcpr case=start
// [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]]\n"ABCCED"\n
// @lcpr case=end

// @lcpr case=start
// [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]]\n"SEE"\n
// @lcpr case=end

// @lcpr case=start
// [["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]]\n"ABCB"\n
// @lcpr case=end

 */

